Use libMesh::BoundingBox, not libMesh::MeshTools::BoundingBox

roystgnr opened the pull request:
This fixes a #1500 bug (w.r.t. efficiency, not necessarily correctness), by using the new create_local_bounding_box() API instead of misusing the old processor_bounding_box() API.  But since we're switching to new APIs anyway, we might as well replace all the other deprecated BoundingBox usage with its newer version.

The header refactor which that enables is significant, but as with any #8730 operation it's possible it will break things downstream which had been inappropriately relying on indirectly included headers.  If there is much breakage, then we might just merge the first two commits of this PR right away but leave the third as a longer-term update.

roystgnr committed:
Include missing libMesh headers

These headers are currently being included indirectly, but subsequent
refactoring in the service of #8730 will change that and we won't want
things to break when it does.

roystgnr committed:
Use BoundingBox, not MeshTools::BoundingBox

This gets rid of several deprecated libMesh API calls.

In particular, it gets rid of a *buggy* API call, where
MultiAppTransfer was trying to get a local bounding box via misuse of
processor_bounding_box() and was inadvertently getting a global
bounding box instead.  This was preventing us from merging
https://github.com/libMesh/libmesh/pull/1303 and it resembles other
buggy calls which have in the past caused #1500 tests to fail.

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/116433/) on 9b9e968 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format f58bdcb697eb84013e0ca291c240425584eaf13f`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9858/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9858/style.patch | git apply -v`

roystgnr committed:
clang-format indentation in forward declarations

roystgnr committed:
clang-format whitespace in code

roystgnr commented:
And by "first two commits" vs "third" I now mean "first two and fifth" vs "third and fourth".

Also, *now* I remember why I was going to build a new clang module, before I got distracted by gcc 7 and Intel 2018...

roystgnr committed:
ImageSampler.C needs MeshTools

And we need to switch to the non-deprecated call here while we're at
it.

Not sure why this previously compiled fine for me but not for Civet.

roystgnr committed:
Fix two more deprecated bounding_box() calls

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/116469/) on 033947a wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9858/site)

roystgnr commented:
Looks like all the app failures (so far) come down to the "mesh.h"->"mesh_base.h" change in MooseMesh.h.  I'll move that to a separate PR.

roystgnr committed:
BoundingBox now only requires bounding_box.h

Or sometimes only a forward declaration, where we can get away with
it.

roystgnr commented:
@jwpeterson, could you take a look at this sometime?  No great urgency, since the bug instance fixed here appears to be an efficiency rather than a correctness issue, but after the fix goes in we can finally merge libMesh/libmesh#1303 and make sure nobody introduces a more serious case of the same bug in the future.
